ltm/agent.py

```diff
@@ -1,102 +1,408 @@
+
+Write JSON in the following format:
+
+{{
+    "queries": array, // An array of strings: 2 descriptive search phrases, one general and one specific
+    "keywords": array // An array of strings: 1 to 3 keywords that can be used to narrow the category of memories that are interesting. 
+}}"""
+
+        # Now create the context for generating the queries
+        context = [make_user_message(prompt.format(user_message=user_message, time=self.now, keywords=self.defined_kws))]
+        all_retrieved_memories = []
+        query_keywords = []
+        while True:
+            print("generating queries")
+            response = ask_llm(context, model=self.model, max_overall_tokens=self.max_prompt_size, cost_callback=cost_cb, temperature=self.temperature)
+
+            try:
+                query_dict = sanitize_and_parse_json(response)
+                query_keywords = [k.lower() for k in query_dict["keywords"]]
+                print(f"Query keywords: {query_keywords}")
+
+                all_retrieved_memories = []
+                for q in query_dict["queries"] + [user_message]:
+                    print(f"Querying with: {q}")
+                    for mem in self.semantic_memory.retrieve(q, k=100):
+                        all_retrieved_memories.append(mem)
+                break
+            except Exception:
+                continue
+
+        # Filter by both relevance and keywords
+        all_keywords = query_keywords
+        relevance_filtered_mems = [x for x in all_retrieved_memories if x.relevance > 0.6] + self.retrieve_from_keywords(all_keywords)
+        keyword_filtered_mems = []
+
+        for m in relevance_filtered_mems:
+            for kw in m.metadata["keywords"]:
+                if kw in all_keywords:
+                    keyword_filtered_mems.append(m)
+                    break
+
+        keyword_filtered_mems.extend(self.retrieve_from_keywords(all_keywords))
+
+        # Spreading activations
+        print(f"Performing spreading activations with {len(keyword_filtered_mems[:10])} memories.")
+        for mem in keyword_filtered_mems[:10]:
+            # print(f"Spreading with: {mem.passage}")
+            for r_mem in self.semantic_memory.retrieve(mem.passage, k=5):
+                if r_mem.relevance > 0.6:
+                    keyword_filtered_mems.append(r_mem)
+
+        # # TODO: Uncomment all this stuff when doing dev stuff
+        # trivia_skip = False
+        # for kw in all_keywords:
+        #     if "trivia" in kw:
+        #         trivia_skip = True
+        #
+        # if trivia_skip:
+        #     llm_filtered_interactions, _ = self.interactions_from_retrieved_memories(keyword_filtered_mems)
+        # else:
+        #     llm_filtered_interactions = self.llm_memory_filter(keyword_filtered_mems, query_dict["queries"], cost_cb)
+
+        # TODO: ....And comment this one out
+        llm_filtered_interactions = self.llm_memory_filter(keyword_filtered_mems, query_dict["queries"], cost_cb)
```

Comment:
Could it be good to make this depend on the debug level, instead of commenting it in/out?